browser/js/global.js

```diff
@@ -675,6 +675,68 @@ window.app = {
 		}, false);
 	}
 
+	// indirect socket to wrap the asyncness around fetching the routetoken from indirection url endpoint
+	global.IndirectSocket = function(uri) {
+		var that = this;
+		this.uri = uri;
+		this.binaryType = '';
+		this.unloading = false;
+		this.readyState = 0; // connecting
+		this.innerSocket = undefined;
+
+		this.onclose = function() {};
+		this.onerror = function () {};
+		this.onmessage = function () {};
+		this.onopen = function () {};
+
+		this.close = function() {
+			this.readyState = 3;
+			this.innerSocket.onerror = function () {};
+			this.innerSocket.onclose = function () {};
+			this.innerSocket.onmessage = function () {};
+			this.innerSocket.close();
+		};
+
+		this.send = function(msg) {
+			this.innerSocket.send(msg);
+		};
+
+		this.setUnloading = function() {
+			this.unloading = true;
+		};
+
+		var http = new XMLHttpRequest();
+		http.open('GET', global.indirectionUrl, true);
+		http.responseType = 'json';
+		http.addEventListener('load', function() {
+			if (this.status === 200) {
+				global.routeToken = http.response.route_token;
+				//add routeToken in uri
+				var startIndex = that.uri.indexOf('&compat');
+				var uri = that.uri.substr(0, startIndex) + encodeURIComponent('&RouteToken=' + global.routeToken) + that.uri.substr(startIndex);
+				that.innerSocket = new WebSocket(uri);
+				that.innerSocket.binaryType = that.binaryType;
+				that.innerSocket.onerror = function() {
+					that.onerror();
+				};
+				that.innerSocket.onclose = function() {
+					that.readyState = 3;
+					that.onclose();
+				};
+				that.innerSocket.onopen = function() {
+					that.readyState = 1;
+					that.onopen();
+				};
+				that.innerSocket.onmessage = function(e) {
+					that.onmessage(e);
```

Comment:
Looks great  I  might be inclined to clone the 'readyState' across both here (each time before the onmessage)  and also in the 'onerror()' callback just before we emit it =)

Reply:
I'd use the that.readyState; rather than '1' in case there is some dis-connecting / other state we want to clone here I think; in general rather than hard-coding it it would be nice to just copy it from the underlying websocket I think (?) then any odd behavior there will be mirrored in our wrapper (?)